File: 208. Implement Trie -Prefix Tree/solution.py
from collections import defaultdict

class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.chars = {}
        self.eow = False

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        if not word[0] in self.chars:
            self.chars[word[0]] = Trie()
            
        if len(word) == 1:
            self.chars[word[0]].eow = True
        else:
            self.chars[word[0]].insert(word[1:])

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        if not word:
            return self.eow
        else:
            if word[0] in self.chars:
                return self.chars[word[0]].search(word[1:])
            else:
                return False

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        if len(prefix) == 1:
            if prefix[0] in self.chars:
                return True
            else:
                return False
        else:
            if prefix[0] in self.chars:
                return self.chars[prefix[0]].startsWith(prefix[1:])
            else:
                return False


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208. Implement Trie -Prefix Tree/solution.java
class Trie {
    private Node node;
    
    /** Initialize your data structure here. */
    public Trie() {
        node = new Node();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        Node ptr = node;
        int index;
        for (int i = 0; i < word.length(); i++) {
            index = ptr.getIndex(word.charAt(i));
            if (ptr.node[index] == null) {
                ptr.node[index] = new Node();
            }
            ptr = ptr.node[index];
        }
        
        ptr.eow = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        Node ptr = node;
        int index;
        for (int i = 0; i < word.length(); i++) {
            index = ptr.getIndex(word.charAt(i));
            if (ptr.node[index] == null) {
                return false;
            }
            ptr = ptr.node[index];
        }
        
        return ptr.eow;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        Node ptr = node;
        int index;
        for (int i = 0; i < prefix.length(); i++) {
            index = ptr.getIndex(prefix.charAt(i));
            if (ptr.node[index] == null) {
                return false;
            }
            ptr = ptr.node[index];
        }
        
        return true;
    }
    
    class Node {
        Node[] node;
        boolean eow;
        
        public Node() {
            node = new Node[26];
            for (int i = 0; i < 26; i++) {
                node[i] = null;
            }
            
            eow = false;
        }
        
        public int getIndex(char c) {
            return (int)c - 97;
        }
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
